Essentials/src/com/earth2me/essentials/EssentialsPluginListener.java

```diff
@@ -25,6 +27,9 @@ public void onPluginEnable(final PluginEnableEvent event) {
         }
         ess.getPermissionsHandler().setUseSuperperms(ess.getSettings().useBukkitPermissions());
         ess.getPermissionsHandler().checkPermissions();
+        ess.getPermissionsHandler().registerContext("ess-afk", player -> Collections.singleton(String.valueOf(ess.getUser(player).isAfk())), () -> ImmutableSet.of("true", "false"));
```

Comment:
These register calls should probably be in `onEnable`??

Unless I'm missing something, they're currently being added on every `PluginEnableEvent`? (this is gonna fire every time a plugin loads?!)

Reply:
No... I think you're right about that. It is definitely going to fire every time a plugin loads. I actually did originally have these in `onEnable`, but the problem is that before `checkPermissions` is run (which is only done here) the `handler` in `PermissionsHandler` is null. It's probably done in this fashion to avoid adding softdepends on every Vault-based permission plugin.

Two possible solutions:
1) From the event, check if the plugin is a known one that supports contexts.
2) Schedule a task to register contexts on the first server tick.

I don't particularly like either solution, but it doesn't seem like Bukkit has any event that gets called after all plugins are loaded. Maybe something else works too, but this is what comes to mind. Any opinions?

Reply:
So I got another idea from Glare (on Discord) that I like slightly better, which is a spin-off of my first solution. Instead of checking from this event, we do actually do it in `onEnable`, since after we've checked that a context supporting permission plugin is loaded, it should be safe to register the contexts since Essentials has a softdepend on it, and thus `checkPermissions` has to have run already and have a non-null handler.

Reply:
You're also already storing a
```java
private Set<ContextCalculator<Player>> contextCalculators
```

Wouldn't be the biggest change to make that a `Map<String, ContextCalculator<Player>>` and just check to see if already registered - just an idea.

However, I'd suggest re-thinking whether that logic really needs to be running every time a plugin enables in the first place...

Reply:
```java
ess.getPermissionsHandler().setUseSuperperms(ess.getSettings().useBukkitPermissions());
ess.getPermissionsHandler().checkPermissions();
ess.getPermissionsHandler().registerContext(...);
```

can all be moved to `onEnable`, you're softdepending on LuckPerms now anyway.

Then just keep a single call to
```java
ess.getPermissionsHandler().checkPermissions();
```
in the plugin enable listener to catch other permission plugins.

The call to
```java
ess.getPermissionsHandler().unregisterContexts();
```

definitely needs to be moved to `onDisable`.

Reply:
Yeah, it wasn't my intention to leave the registrations there either. It was just an easy place to put them for now while testing everything. They can be registered from anywhere though, whether that's primarily in `onEnable`, or a `registerContexts` method. For the latter, we probably want to consider if all contexts will always be enabled all at once, or if there are contexts which may need to be initialized elsewhere (because the calculator function needs to be defined in scope of whatever it's using, although most of the time that would probably be the `Essentials` instance).